### PR TITLE
make in_pure_callback state thread-local

### PR DESCRIPTION
**Pull request by vtjnash**

### CODE REVIEW

**Comment by tkelman** [comment]
what is this fixing, why is it necessary, add a test.

File: src/julia_threads.h [hunk 1]
```diff
@@ -107,6 +107,10 @@ typedef struct _jl_tls_states_t {
     pthread_t system_id;
     void *signal_stack;
 #endif
+    // execution of certain certain unpure
```
Comment (tkelman) [inline src/julia_threads.h]: impure

**Comment by yuyichao** [comment]
A few other global variables that should be also thread local.

* `fl_current_module_counter::fallback_counter`
* `jl_toplevel_eval_in_warn::jl_warn_on_eval`
* `jl_filename`
* `jl_lineno`
* `jl_eval_module_expr::module_stack`
* `jl_eval_module_expr::initialized`
* `jl_eval_module_expr::outermost`

These are on my low priority todo list since they are hard to trigger and test.

There's also a script [here](https://github.com/yuyichao/explore/blob/master/julia/check_thread_gv/parseobjs2.jl) which list all the global variables but I haven't finished the white list yet.....

**Comment by vtjnash** [comment]
I think most of those will end up getting protected by the global "toplevel lock". Making a new module on a thread may be valid, but I'm not sure it should be allowed.

**Comment by ViralBShah** [comment]
Does this fix any known issue?

**Comment by vtjnash** [comment]
Staged functions are still a bit race-y with multithreading. I'm almost done with a PR to fix that. This was just an easy no-brainer fix.

**Comment by vtjnash** [comment]
I can't reliably test this right now as the code-path is still broken, this is just a minor piece towards fixing it.

**Comment by tkelman** [comment]
if this doesn't yet fix an actual problem then it's not getting backported

**Comment by vtjnash** [comment]
sgtm. I'm very much in favor of making the v0.6 release around Dec-Jan that brings in lots of threading fixes + 265 + fixed type system.

**Comment by timholy** [comment]
With that schedule, 0.6 might get released before 0.5 :smile:.

**Pull request merged**